guice-jpa/src/main/java/com/maxifier/guice/jpa/UnitOfWork.java

```diff
@@ -0,0 +1,160 @@
+package com.maxifier.guice.jpa;
+
+import com.maxifier.guice.jpa.DB.Transaction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.EntityTransaction;
+
+/**
+ * Class used for database context handling.
+ * <p>Class uses thread-local state to handle database context propagation.</p>
+ * <p>Database context can be handled manually. Use manual handling with caution, mismatch between
+ * {@code begin()} and {@code end()} calls cause huge memory leaks.</p>
+ * <p>Usage:</p>
+ * <pre>
+ * class Worker {
+ *     &#64;Inject EntityManager em;
+ *
+ *     void doWork() {
+ *         UnitOfWork.begin();
+ *         try {
+ *             em.find(Foo.class, 7);
+ *         } finally {
+ *             UnitOfWork.end();
+ *         }
+ *     }
+ * }
+ * </pre>
+ * <p>Preferred for use in tests. Use {@link DBEntityManagerProvider} to create proper EM.</p>
+ *
+ * @author Konstantin Lyamshin (2015-11-09 18:22)
+ */
+public final class UnitOfWork {
+    private static final Logger logger = LoggerFactory.getLogger(UnitOfWork.class);
+    private static final ThreadLocal<UnitOfWork> current = new ThreadLocal<UnitOfWork>();
+    private final UnitOfWork previous;
+    private EntityManager entityManager;
+    private boolean startTransaction;
+
+    /**
+     * Starts a new UnitOfWork for the current thread.
+     * <p>UnitOfWork holds actual database context.</p>
+     * <p>Always creates new UnitOfWork like in {@link Transaction#REQUIRES_NEW} mode. But don't start transaction
+     * like in {@link Transaction#NOT_REQUIRED} mode. All transactions should be handled manually.</p>
+     */
+    public static void begin() {
+        new UnitOfWork();
+    }
+
+    /**
+     * Finishes current UnitOfWork.
+     * <p>Current UnitOfWork must be previously started by {@link #begin()}.</p>
+     */
+    public static void end() {
+        UnitOfWork context = current.get();
+        if (context != null) {
+            context.releaseConnection();
+        } else {
+            logger.error("Corrupted UnitOfWork call stack", new IllegalStateException());
+        }
+    }
+
+    @Nullable
+    static UnitOfWork get() {
+        return current.get();
+    }
+
+    UnitOfWork() {
+        this.previous = current.get();
+        current.set(this);
```

Comment:
This seems to be hacky and too implicit. Should we instead move this out of constructor?